pkg/coalescer/coalescer.go

```diff
@@ -0,0 +1,59 @@
+package coalescer
+
+import (
+	"sync"
+)
+
+// Coalescer contains a Run function that will run a specified function.
+// If the function is already being run in a separate thread, the Coalescer
+// will simply wait until the function finishes and return nil
+type Coalescer interface {
+	Run() error
+}
+
+// NewCoalescer returns a new Coalescer with the given function set as the runFunc.
+func NewCoalescer(f func() error) Coalescer {
+	return &coalescer{runFunc: f, loader: sync.NewCond(&sync.Mutex{})}
+}
+
+type coalescer struct {
+	sync.Mutex
+	loading bool
+	loader  *sync.Cond
+	runFunc func() error
+}
+
+// Run the runFunc. If the runFunc is currently being run in another thread, wait
+// and return nil when the the other thread finishes.
+func (c *coalescer) Run() error {
+	c.Lock()
+	if c.loading {
+		// someone else is running it, so we wait
+		c.loader.L.Lock()
+		c.Unlock()
+		c.loader.Wait()
+		c.loader.L.Unlock()
+		return nil
+	}
+	// nobody else is loading
+	c.loading = true
+	c.Unlock()
+
+	// reload the registry
+	err := c.runFunc()
+	if err != nil {
+		c.loader.L.Lock()
+		c.loader.Broadcast()
+		c.loader.L.Unlock()
+		return err
+	}
+
+	// inform the waiters that we are done
+	c.Lock()
+	c.loading = false
+	c.loader.L.Lock()
+	c.loader.Broadcast()
+	c.loader.L.Unlock()
+	c.Unlock()
+	return nil
```

Comment:
Looking at the resolver, it seems subsequent invocations don't actually need to wait for the first, is that right?
- Yes:
```go
// sync/atomic
if atomic.CompareAndSwapUint32(&loading, 0, 1) {
    err = f()
    atomic.StoreUint32(&loading, 0)
}
```
(there may be an equivalent in the stdlib, but this works)
- No:
```go
// sync.Once
once.Do(func() { err = f() })
```

Reply:
The resolver itself should work fine even if there are multiple invocations running at the same time and it doesn't actually need to wait. However, I prefer having the call to coalescer to be blocking, so that if a function calls it and the reload is going on, the caller waits until an up to date version of the resolver is ready.